miden-tx/src/testing/mock_chain/mod.rs

```diff
@@ -0,0 +1,796 @@
+    created_nullifiers: Vec<Nullifier>,
+
+    /// Transaction IDs added to the block.
+    included_transactions: Vec<(TransactionId, AccountId)>,
+}
+
+impl PendingObjects {
+    pub fn new() -> PendingObjects {
+        PendingObjects {
+            updated_accounts: vec![],
+            output_note_batches: vec![],
+            created_nullifiers: vec![],
+            included_transactions: vec![],
+        }
+    }
+
+    /// Creates a [BlockNoteTree] tree from the `notes`.
+    ///
+    /// The root of the tree is a commitment to all notes created in the block. The commitment
+    /// is not for all fields of the [Note] struct, but only for note metadata + core fields of
+    /// a note (i.e., vault, inputs, script, and serial number).
+    pub fn build_notes_tree(&self) -> BlockNoteTree {
+        let entries =
+            self.output_note_batches.iter().enumerate().flat_map(|(batch_index, batch)| {
+                batch.iter().enumerate().map(move |(note_index, note)| {
+                    (
+                        BlockNoteIndex::new(batch_index, note_index).unwrap(),
+                        note.id(),
+                        *note.metadata(),
+                    )
+                })
+            });
+
+        BlockNoteTree::with_entries(entries).unwrap()
+    }
+}
+
+#[derive(Debug)]
+pub enum MockError {
+    DuplicatedNullifier,
+    DuplicatedNote,
+}
+
+impl fmt::Display for MockError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for MockError {}
+
+// MOCK CHAIN
+// ================================================================================================
+
+/// [MockChain] simulates a simplified blockchain environment for testing purposes.
+/// It allows to create and manage accounts, mint assets, execute transactions, and apply state
+/// updates.
+///
+/// This struct is designed to mock transaction workflows, asset transfers, and
+/// note creation in a test setting. Once entities are set up, [TransactionContextBuilder] objects
+/// can be obtained in order to execute transactions accordingly.
+///
+/// # Examples
+///
+/// ## Create mock objects
+/// ```
+/// let mut mock_chain = MockChain::default();
+/// let faucet = mock_chain.add_new_faucet(Auth::BasicAuth, "USDT", 100_000);  // Create a USDT faucet
+/// let asset = faucet.mint(1000);  
+/// let note = mock_chain.add_p2id_note(asset, sender...);
+/// let sender = mock_chain.add_new_wallet(Auth::BasicAuth, vec![]);  
+///
+/// mock_chain.build_tx_context(sender.id(), &[note.id()], &[]).build().execute()
+/// ```
+///
+/// ## Executing a Simple Transaction
+///
+/// NOTE: Transaction script is defaulted to either [DEFAULT_AUTH_SCRIPT] if the account includes
+/// an authenticator.
+///
+/// ```
+/// let mut mock_chain = MockChain::default();
+/// let sender = mock_chain.add_new_wallet(Auth::BasicAuth, vec![asset]);  // Add a wallet with assets
```

Comment:
Just FYI: This is no longer allowed in the new AccountBuilder after #952 since new accounts are not allowed to have assets. This can probably just be changed to be an existing account.

Reply:
Removed the ability of setting assets and changed the examples